onnxruntime/core/graph/graph.cc

```diff
@@ -3030,6 +3030,96 @@ ONNX_NAMESPACE::GraphProto Graph::ToGraphProto() const {
   return result;
 }
 
+ONNX_NAMESPACE::GraphProto Graph::ToGraphProtoWithExternalInitializers(const std::string& external_file_name) const {
+  // Sparse tensors are not supported yet.
+  ORT_ENFORCE(sparse_tensor_names_.empty());
+
+  GraphProto graph_proto;
+  graph_proto.set_name(Name());
+  graph_proto.set_doc_string(Description());
+
+  std::ofstream output_stream(external_file_name, std::ofstream::out | std::ofstream::binary);
+  if (!output_stream.is_open())
+    throw std::runtime_error("Cannot open file " + external_file_name);
+
+  // Add all initializers to the graph proto dumping their content to the external file.
+  int64_t file_offset = 0;
+  for (const auto& iter : GetAllInitializedTensors()) {
+    TensorProto* output_proto = graph_proto.add_initializer();
+    const TensorProto* input_proto = iter.second;
+
+    size_t tensor_bytes_size = 0;
+    std::unique_ptr<uint8_t[]> raw_data;
+    ORT_THROW_IF_ERROR(utils::UnpackInitializerData(*input_proto, Path(), raw_data, tensor_bytes_size));
+
+    // Only place floating-point tensors in the external buffer.
+    // This makes sure that shape-constants used by Reshape and Transpose are included in the onnx file.
+    // Without these the onnx shape and type inference pass does not work.
+    if (input_proto->data_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT &&
+        input_proto->data_type() != ONNX_NAMESPACE::TensorProto_DataType_DOUBLE &&
+        input_proto->data_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT16 &&
+        input_proto->data_type() != ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16) {
+      *output_proto = *input_proto;
```

Comment:
May be useful to have a minimum size before storing externally, or to just use a minimum size as that would also rule out all the shape constants. This is what the ONNX script that converts to using external files does. 

[https://github.com/onnx/onnx/blob/859ca2d9840e23fcdac91df2469d0ac37c0a4187/onnx/external_data_helper.py#L113](https://github.com/onnx/onnx/blob/859ca2d9840e23fcdac91df2469d0ac37c0a4187/onnx/external_data_helper.py#L113)
    size_threshold: Threshold for size of data. Only when tensor's data is >= the size_threshold

Reply:
Thanks, I was not aware of the python script. I'll add a threshold parameter here.